bindings/rust/s2n-tls/src/callbacks/pkey.rs

```diff
@@ -0,0 +1,371 @@
+                    Poll::Ready(Err(error::Error::application(
+                        "missing pkey operation".into(),
+                    )))
+                }
+            }
+        }
+
+        struct TestPkeyCallback(Counter);
+        impl PrivateKeyCallback for TestPkeyCallback {
+            fn handle_operation(
+                &self,
+                _conn: &mut connection::Connection,
+                op: PrivateKeyOperation,
+            ) -> Result<Option<Pin<Box<dyn ConnectionFuture>>>, error::Error> {
+                self.0.increment();
+                let future = TestPkeyFuture {
+                    counter: 0,
+                    op: Some(op),
+                };
+                Ok(Some(Box::pin(future)))
+            }
+        }
+
+        let (waker, wake_count) = new_count_waker();
+        let counter = testing::Counter::default();
+        let callback = TestPkeyCallback(counter.clone());
+        let pair = new_pair(callback, waker)?;
+
+        assert_eq!(counter.count(), 0);
+        assert_eq!(wake_count, 0);
+        poll_tls_pair(pair);
+        assert_eq!(counter.count(), 1);
+        assert_eq!(wake_count, POLL_COUNT);
+
+        Ok(())
+    }
+
+    #[test]
+    fn sync_failure() -> Result<(), Error> {
+        const ERROR: &str = "sync_failure error";
+
+        struct TestPkeyCallback(Counter);
+        impl PrivateKeyCallback for TestPkeyCallback {
+            fn handle_operation(
+                &self,
+                _conn: &mut connection::Connection,
+                _op: PrivateKeyOperation,
+            ) -> Result<Option<Pin<Box<dyn ConnectionFuture>>>, error::Error> {
+                self.0.increment();
+                Err(testing::test_error(ERROR))
+            }
+        }
+
+        let (waker, wake_count) = new_count_waker();
+        let counter = testing::Counter::default();
+        let callback = TestPkeyCallback(counter.clone());
+        let pair = new_pair(callback, waker)?;
+
+        assert_eq!(counter.count(), 0);
+        assert_eq!(wake_count, 0);
+        let result = poll_tls_pair_result(pair);
+        assert_eq!(counter.count(), 1);
+        assert_eq!(wake_count, 0);
+
+        match result {
+            Ok(_) => panic!("Handshake unexpectedly succeeded"),
+            Err(e) => testing::assert_test_error(e, ERROR),
+        };
+        Ok(())
+    }
+
+    #[test]
+    fn async_failure() -> Result<(), Error> {
+        const POLL_COUNT: usize = 10;
+        const ERROR: &str = "async_failure error";
+
+        struct TestPkeyFuture {
+            counter: usize,
+            _op: PrivateKeyOperation,
+        }
+        impl ConnectionFuture for TestPkeyFuture {
+            fn poll(
+                mut self: Pin<&mut Self>,
+                conn: &mut connection::Connection,
+                _ctx: &mut core::task::Context,
+            ) -> Poll<Result<(), error::Error>> {
+                conn.waker().expect("Uncaught missing waker").wake_by_ref();
```

Comment:
I think it would be better to go through the `ctx.waker()` rather than the connection. I'd actually like to get rid of getting/setting a waker on the connection in the public API, since it requires cloning it to do that safely in the general case (see #3748)